fuzz: split `coinselection` harness

brunoerg opened the pull request:
This PR splits the `coinselection` fuzz harness into 3 targets (`coinselection_bnb`, `coinselection_knapsack`, `coinselection_srd`). The goal is to be able to fuzz each algorithm separately (to avoid performance issues) and also all of them together.

brunoerg committed:
fuzz: split `coinselection` harness

brunoerg commented on src/wallet/test/fuzz/coinselection.cpp:
```diff
@@ -249,69 +256,67 @@ FUZZ_TARGET(coinselection)
-        assert(result_knapsack->GetSelectedValue() >= target);
-        assert(result_knapsack->GetWeight() <= max_selection_weight);
-        result_knapsack->RecalculateWaste(coin_params.min_viable_change, coin_params.m_cost_of_change, coin_params.m_change_fee);
-        (void)result_knapsack->GetShuffledInputVector();
```
There is no need to remove it, will put back. Probably missed it during the rework. Thanks.

brunoerg replied:
Done

brunoerg commented on src/wallet/test/fuzz/coinselection.cpp:
```diff
@@ -216,10 +216,17 @@ FUZZ_TARGET(coin_grinder_is_optimal)
+};
+
+template<CoinSelectionAlgorithm Algorithm>
+void FuzzCoinSelectionAlgorithm(const uint8_t* buffer, size_t size) {
```
Done

brunoerg commented on src/wallet/test/fuzz/coinselection.cpp:
```diff
@@ -216,10 +216,17 @@ FUZZ_TARGET(coin_grinder_is_optimal)
+    BNB,
+    SRD,
+    KNAPSACK,
+    ALL
```
I just removed the `ALL` option. I don't see anything that needs all of them combined.

brunoerg commented:
Force-pushed addressing https://github.com/bitcoin/bitcoin/pull/31870#discussion_r1957893638, https://github.com/bitcoin/bitcoin/pull/31870#discussion_r1957883443 and https://github.com/bitcoin/bitcoin/pull/31870#discussion_r1957883107. Thanks, @maflcko.

brunoerg commented on src/wallet/test/fuzz/coinselection.cpp:
```diff
@@ -331,4 +343,16 @@ FUZZ_TARGET(coinselection)
+}
+
+FUZZ_TARGET(coinselection_knapsack) {
+    FuzzCoinSelectionAlgorithm<CoinSelectionAlgorithm::KNAPSACK>(std::span<const uint8_t>(buffer.data(), buffer.size()));
```
yes, done.

zaidmstrr reviewed (commented):
Hey, I'm reviewing this PR, as the description says that we can test each target separately, (`coinselection_bnb`, `coinselection_knapsack`, `coinselection_srd`) which is fine, but for the combined test I can't find any fuzz target there like `coinselection`.

brunoerg commented:
> Hey, I'm reviewing this PR, as the description says that we can test each target separately, (`coinselection_bnb`, `coinselection_knapsack`, `coinselection_srd`) which is fine, but for the combined test I can't find any fuzz target there like `coinselection`.

I just updated the description. I removed the `coinselection` target, so there are 3 ones.

zaidmstrr reviewed (commented):
Tested ACK [9b293d0](https://github.com/bitcoin/bitcoin/pull/31870/commits/9b293d031c941dc6af0c4274e43388c57c702047) 
The idea of fuzzing each algorithm independently seems good. I manually reviewed the new code changes with the previous ones to check for any missing statements. I also tested the code by manually inserting assert statements in between to check whether the desired behaviour was occurring or not.

marcofleon reviewed (commented):
Code review ACK 9b293d031c941dc6af0c4274e43388c57c702047

Knapsack is quite a bit slower than the other two, so separating them out is probably a good idea.

Coverage for the three targets: 
[coinselection_bnb](https://marcofleon.github.io/coverage/coinselection_bnb/)
[coinselection_srd](https://marcofleon.github.io/coverage/coinselection_srd/)
[coinselection_knapsack](https://marcofleon.github.io/coverage/coinselection_knapsack/)

brunoerg commented on src/wallet/test/fuzz/coinselection.cpp:
```diff
@@ -249,66 +256,72 @@ FUZZ_TARGET(coinselection)
+            // If the total balance is sufficient for the target and we are not using
+            // effective values, Knapsack should always find a solution (unless the selection exceeded the max tx weight).
+            if (total_balance >= target && subtract_fee_outputs && !HasErrorMsg(result_knapsack)) {
+                assert(result_knapsack);
```
Removing?

marcofleon replied:
Good catch. Took me a bit to see, but yeah this check should be outside of the inner `if` statement.

brunoerg replied:
Ah, good catch! Gonna fix it, thanks.

brunoerg replied:
Done.

brunoerg commented:
Force-pushed addressing https://github.com/bitcoin/bitcoin/pull/31870#discussion_r1965528993

janb84 commented:
Tested ACK [ba82240](https://github.com/bitcoin/bitcoin/pull/31870/commits/ba82240553ddd534287845e10bc76b46b45329fe)

The change to fuzz each algorithm independently seems an okay idea. The implementation/separation in the code also makes it easier to read, review, and reason about.

- Built and ran fuzzers, minimum of 15 minutes each. 
- Reviewed code changes line-by-line
- Tried to compile CodeCoverage of new Fuzz targets but I failed at that (not a code issue) 

I also wrote a [review note](https://github.com/janb84/Bitcoin-Core_review-notes/blob/main/reviews/PR-%2331870.md) on the PR to document my learning journey

marcofleon reviewed (commented):
reACK ba82240553ddd534287845e10bc76b46b45329fe

Only thing that changed since last review is the fix to https://github.com/bitcoin/bitcoin/pull/31870#discussion_r1965528993. The order of the two checks has been switched but afaict it shouldn't matter.

yancyribbens commented:
Concept ACK.  I'm curious if there should be any doc updates accompanying this change.

zaidmstrr reviewed (commented):
reACK [ba82240](https://github.com/bitcoin/bitcoin/pull/31870/commits/ba82240553ddd534287845e10bc76b46b45329fe)
As the previous issue raised by maflcko is resolved.

fanquake commented:
cc @dergoegge
